ADR/00xx-component-references.md

```diff
@@ -0,0 +1,147 @@
+
+In theory **any combination of the cases above** could be present in an app. They could have one common parent image, many operand images that depend on that, and one or more operator bundles which depend on those.
+
+Today, users work around how complicated it is to manage digests themselves by instead using *floating tags*, which have the benefit of being easy to use - no need to update - but have the problem of being unclear. It's not exactly clear what you're building against if you refer to it by tag. Potentially insecure. We want to make it easy for users to do better.
+
+## Decision
+
+### API
+
+* Introduce a new CustomResource called `ComponentReference` that lets one Component declare that another Component **embeds a reference** to it.
+  * A `ComponentReference` spec contains two fields: a referent, and a referer
+  * [integration-service] handles the testing and promotion of all referents in a special way.
+  * A new set of functionality for [build-service] uses the `ComponentReference` links to propose updates to users git repositories.
+* Introduce a new convention for PRs that lets one PR declare that it depends on another PR.
+* Important to understand: the declared reference embeddings as defined in the `ComponentReferences` and PR dependencies are two different things.
+  * `ComponentReferences` are declarations about which Components we should expect to see PR dependencies for.
+    * You say in english: “Component A references Component B.”
+  * PR dependencies can be supplied by the user without any declared `ComponentReferences` on the respective Components.
+    * You say in english: “PR #1 depends on PR #2.”
+
+### Integration-service
+
+* When [integration-service] notices a build of a *referent* Component (where another Component declares that it embeds a reference to this one), it always skips all testing (both pre-merge and post-merge). It always promotes the image to the global candidate list when the PR is merged, but it does not promote to dev environment (post-merge) and it does not create Releases (post-merge).
+  * Builds of *referent* Components never directly trigger tests, never trigger promotion, or Releases.
+* When [integration-service] notices a build of a *referer* Component (one that declares that it embeds references to others), it does testing as normal, promotes to global candidate list as normal, and it promotes to dev env as normal, and it creates Releases as normal.
+* When [integration-service] notices a build of a Component that references no other components and no other components refer to it - then it proceeds like normal, runs tests, promotes, deploys, releases, etc.
+
+### Build-service
+
+* [build-service] will propagate the digest from one Component to another as a PR – following the declared `ComponentReferences`. This is a bit like renovatebot wrapped in a controller. The [build-service] acts when it sees that a build pipelinerun has completed successfully for a *referent* image.
```

Comment:
what are the mechanics for **propagate the digest from one Component to another as a PR**?

Reply:
It looks like a recursive `sed` invocation in the repo, submitting as a pull request. Replace all instance of the old digest with the new digest.

Reply:
How does the Build Service know what the old digest is? Via the Component CR?